File: product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        var arrProduct = 1;
        var countZeroes = 0;
        for (var a: nums) {
            if (a != 0)
                arrProduct *= a;
            else {
                countZeroes++;
            }
        }
        var ans = new int[nums.length];
        Arrays.fill(ans, countZeroes > 0 ? 0 : arrProduct);
        for(int i=0; i<nums.length; i++) {
            if (nums[i] == 0 && countZeroes == 1) {
                ans[i] = arrProduct;
            } else if(nums[i] != 0) {
                ans[i] /= nums[i];    
            }
        }
        return ans;
    }
}

File: product-of-array-except-self/product-of-array-except-self.ts
let print = console.log;

function productExceptSelf(nums: number[]): number[] {
    let l = nums.length;
    let leftP = new Array(l), rightP = new Array(l);
    leftP[0] = 1, rightP[l - 1] = 1;
    
    for(let i = 1; i < l; i++) 
        leftP[i] = leftP[i - 1] * nums[i - 1];
    
    for(let i = l - 2; i >= 0; i--)
        rightP[i] = rightP[i + 1] * nums[i + 1];
    
    for(let i = 0; i < l; i++) 
        rightP[i] = leftP[i] * rightP[i];
    
    return rightP;
};